README.md
# Image Classifier made with CNN
This is a generic Classifier made with Keras and Tensorflow.

NN arch used here:
- Own simply arch
- Resnet 50

## How to run 
Training with

```python
python classification.py
```

Test with
```python
python classification.py <weights.hdf5>
```


CHANGELOG.md
# baseline
- weights-56-0.77.hdf5
- [ balanced-score 0.62 ]
- [ resnet ]
[[   5    2    1   15    0    7    0    2    2    4    3    7    0    0    1    0]
 [   0  312   17    1    0    0    0    0    1   11    2    0    0    0    0    0]
 [   0   19 1404    6    0    0    0    0    2   10   13    6    0    1    1    0]
 [   2    3   16  139    0   11    1    0    3    9    9    7    0    0    1    0]
 [   0    0    0    1  132    0    6   15    4    0    0    0    0    0    0    0]
 [   2    1   11   52    0   49    3    0    6    5    3    1    0    1    1    0]
 [   2    8   11   10    5    7  125   35   19    2    2    9    2   11    1    0]
 [   0    1    2    0   14    0   37  222    7    1    0    7    0    2    0    0]
 [   1    4   31   36    1   15    4    3   87    7   19   25    0    2    3    0]
 [   1    3    3   13    0    1    0    0    0  106    0    4    0    0    3    0]
 [   0    1   71    9    0    0    0    0    3    9   36   13    0    0    0    0]
 [   0    1   30   13    0    3    0    0   15   12   23  107    0    0    3    0]
 [   0    0    0    0   12    0    8    3    0    0    0    0  175    5    3    1]
 [   0    2    0    1    1    0    9    4    3    0    0    0    0  107    0    0]
 [   3    1    5    2    0    4    3    0    2    4    2    4    0    4   39    3]
 [   0    1    5    1    0    0    1    1    6    0    1    2    1    0    2   44]]


# Class 02 balanced 
- weights-02-0.68.hdf5
- [balanced-score 0.61]
- [ resnet ]
[[   3    1    0   12    0    8    0    2    5    5    5    5    0    0    3    0]
 [   0  310    6    1    0    0    0    0    3   18    4    1    0    1    0    0]
 [   0   53 1232   11    0    3    1    1   33   47   67    8    0    0    6    0]
 [   0    5    7  129    0   11    2    0    7   19   14    6    0    0    1    0]
 [   0    0    0    1  128    0    6   21    2    0    0    0    0    0    0    0]
 [   0    2    4   38    0   63    3    0   10    5    2    4    0    2    2    0]
 [   0    6    2    8    4    5  129   50    9    2    3   12    1   15    3    0]
 [   0    0    0    0   13    1   16  252    7    2    0    0    0    2    0    0]
 [   0    5   11   23    1   16    5    3   97   12   37   19    0    2    6    1]
 [   0    3    0   10    0    0    0    0    1  115    3    2    0    0    0    0]
 [   0    2   38    8    0    1    0    0    5   13   55   20    0    0    0    0]
 [   0    2   13   12    0    3    0    1   32   27   38   75    0    2    2    0]
 [   0    0    0    0    9    0   17    3    0    0    0    0  166    9    1    2]
 [   0    1    0    0    1    0   12    7    1    1    0    0    0  104    0    0]
 [   0    2    2    1    0    2    2    0    6    8    3    7    0    4   37    2]
 [   0    2    1    5    0    0    6    1    3    3    4    1    0    2    1   36]]

layer_rotation_control.py
'''
Code for applying Layca on SGD, Adam, RMSprop and Adagrad.
Source: code is based on keras' implementation of the original optimization methods.
'''

from keras.optimizers import Optimizer
import keras.backend as K
from keras.legacy import interfaces

import numpy as np

def norm(w):
    '''
    computes frobenius norm of a tensor w
    '''
    return K.sqrt(K.sum(w*w))

def layca(p, step, lr):
    '''
    Core operations of layca.
    Takes the current parameters and the step computed by an optimizer, and 
         - projects and normalizes the step such that the rotation operated on the layer's weights is controlled
         - after the step has been taken, recovers initial norms of the parameters
         
    !!! 
    only kernels are optimized. Biases and batchnorm pararmeters are left unchanged. This did not affect performance in our experiments. 
    One can decide to train them anyway (without layca operations) by changing last line to:
    return p  - lr * step
    !!!
    '''
    if 'kernel' in p.name: # only kernels are optimized when using Layca (and not biases and batchnorm parameters)
        # projecting step on tangent space of sphere -> orthogonal to the parameters p
        initial_norm = norm(p)
        step = step - (K.sum(step * p))* p / initial_norm**2

        # normalizing step size (with special attention to numerical problems)
        step = K.switch(norm(step)<= K.epsilon(), lambda: K.zeros_like(step), lambda: step/ (norm(step)) * initial_norm)
        
        # applying step
        new_p =  p - lr * step

        # recovering norm of the parameter from before the update
        new_p = new_p / norm(new_p) * initial_norm
        return new_p
    else:
        return p # - lr * step  # uncomment to train biases and batchnorm parameters (without layca)
            
class SGD(Optimizer):
    """Stochastic gradient descent optimizer.
    Includes support for momentum,
    learning rate decay, and Nesterov momentum.
    # Arguments
        lr: float >= 0. Learning rate.
        momentum: float >= 0. Parameter that accelerates SGD
            in the relevant direction and dampens oscillations.
        decay: float >= 0. Learning rate decay over each update.
        nesterov: boolean. Whether to apply Nesterov momentum.
        layca: boolean, wether to apply layca or not
    """

    def __init__(self, lr=0.01, momentum=0., decay=0.,
                 nesterov=False, layca = False,**kwargs):
        super().__init__(**kwargs)
        with K.name_scope(self.__class__.__name__):
            self.iterations = K.variable(0, dtype='int64', name='iterations')
            self.lr = K.variable(lr, name='lr')
            self.momentum = K.variable(momentum, name='momentum')
            self.decay = K.variable(decay, name='decay')
        self.initial_decay = decay
        self.nesterov = nesterov
        
        self.layca = layca

    @interfaces.legacy_get_updates_support
    def get_updates(self, loss, params):        
        grads = self.get_gradients(loss, params)
        self.updates = [K.update_add(self.iterations, 1)]

        lr = self.lr
        if self.initial_decay > 0:
            lr *= (1. / (1. + self.decay * K.cast(self.iterations,
                                                  K.dtype(self.decay))))
        # momentum
        shapes = [K.int_shape(p) for p in params]
        moments = [K.zeros(shape) for shape in shapes]
        self.weights = [self.iterations] + moments
        for p, g, m in zip(params, grads, moments):  
            v = self.momentum * m - lr * g  # velocity
            self.updates.append(K.update(m, v))
             
            if self.nesterov:
                step =  self.momentum * v - lr * g
            else:
                step =  v
            
            if self.layca:
                new_p = layca(p, -step, lr)
            else:
                new_p =  p + step

            # Apply constraints.
            if getattr(p, 'constraint', None) is not None:
                new_p = p.constraint(new_p)

            self.updates.append(K.update(p, new_p))
        return self.updates

    def get_config(self):
        config = {'lr': float(K.get_value(self.lr)),
                  'momentum': float(K.get_value(self.momentum)),
                  'decay': float(K.get_value(self.decay)),
                  'nesterov': self.nesterov,
                  'layca':self.layca}
        base_config = super().get_config()
        return dict(list(base_config.items()) + list(config.items()))

class RMSprop(Optimizer):
    """RMSProp optimizer.
    It is recommended to leave the parameters of this optimizer
    at their default values
    (except the learning rate, which can be freely tuned).
    This optimizer is usually a good choice for recurrent
    neural networks.
    # Arguments
        lr: float >= 0. Learning rate.
        rho: float >= 0.
        epsilon: float >= 0. Fuzz factor. If `None`, defaults to `K.epsilon()`.
        decay: float >= 0. Learning rate decay over each update.
        layca: boolean, wether to apply layca or not
    # References
        - [rmsprop: Divide the gradient by a running average of its recent magnitude](http://www.cs.toronto.edu/~tijmen/csc321/slides/lecture_slides_lec6.pdf)
    """

    def __init__(self, lr=0.001, rho=0.9, epsilon=None, decay=0., layca = False,
                 **kwargs):
        super(RMSprop, self).__init__(**kwargs)
        with K.name_scope(self.__class__.__name__):
            self.lr = K.variable(lr, name='lr')
            self.rho = K.variable(rho, name='rho')
            self.decay = K.variable(decay, name='decay')
            self.iterations = K.variable(0, dtype='int64', name='iterations')
        if epsilon is None:
            epsilon = K.epsilon()
        self.epsilon = epsilon
        self.initial_decay = decay
        
        self.layca = layca

    @interfaces.legacy_get_updates_support
    def get_updates(self, loss, params):        
        grads = self.get_gradients(loss, params)
        accumulators = [K.zeros(K.int_shape(p), dtype=K.dtype(p)) for p in params]
        self.weights = accumulators
        self.updates = [K.update_add(self.iterations, 1)]

        lr = self.lr
        if self.initial_decay > 0:
            lr *= (1. / (1. + self.decay * K.cast(self.iterations,
                                                  K.dtype(self.decay))))

        for p, g, a in zip(params, grads, accumulators):
            # update accumulator
            new_a = self.rho * a + (1. - self.rho) * K.square(g)
            self.updates.append(K.update(a, new_a))
            #new_p = p - lr * g / (K.sqrt(new_a) + self.epsilon)
            step = lr * g / (K.sqrt(new_a) + self.epsilon)
            
            if self.layca:
                new_p = layca(p, step, lr)
            else:
                new_p =  p - step
            
            # Apply constraints.
            if getattr(p, 'constraint', None) is not None:
                new_p = p.constraint(new_p)

            self.updates.append(K.update(p, new_p))
        return self.updates

    def get_config(self):
        config = {'lr': float(K.get_value(self.lr)),
                  'rho': float(K.get_value(self.rho)),
                  'decay': float(K.get_value(self.decay)),
                  'epsilon': self.epsilon,
                  'layca':self.layca}
        base_config = super(RMSprop, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))
    
class Adam(Optimizer):
    """Adam optimizer.
    Default parameters follow those provided in the original paper.
    # Arguments
        lr: float >= 0. Learning rate.
        beta_1: float, 0 < beta < 1. Generally close to 1.
        beta_2: float, 0 < beta < 1. Generally close to 1.
        epsilon: float >= 0. Fuzz factor. If `None`, defaults to `K.epsilon()`.
        decay: float >= 0. Learning rate decay over each update.
        layca: boolean, wether to apply layca or not
        amsgrad: boolean. Whether to apply the AMSGrad variant of this
            algorithm from the paper "On the Convergence of Adam and
            Beyond".
    # References
        - [Adam - A Method for Stochastic Optimization](http://arxiv.org/abs/1412.6980v8)
        - [On the Convergence of Adam and Beyond](https://openreview.net/forum?id=ryQu7f-RZ)
    """

    def __init__(self, lr=0.001, beta_1=0.9, beta_2=0.999, layca = False,
                 epsilon=None, decay=0., amsgrad=False, **kwargs):
        super(Adam, self).__init__(**kwargs)
        with K.name_scope(self.__class__.__name__):
            self.iterations = K.variable(0, dtype='int64', name='iterations')
            self.lr = K.variable(lr, name='lr')
            self.beta_1 = K.variable(beta_1, name='beta_1')
            self.beta_2 = K.variable(beta_2, name='beta_2')
            self.decay = K.variable(decay, name='decay')
        if epsilon is None:
            epsilon = K.epsilon()
        self.epsilon = epsilon
        self.initial_decay = decay
        self.amsgrad = amsgrad
        
        self.layca = layca
    
    @interfaces.legacy_get_updates_support
    def get_updates(self, loss, params):
        grads = self.get_gradients(loss, params)
        self.updates = [K.update_add(self.iterations, 1)]

        lr = self.lr
        if self.initial_decay > 0:
            lr *= (1. / (1. + self.decay * K.cast(self.iterations,
                                                  K.dtype(self.decay))))

        t = K.cast(self.iterations, K.floatx()) + 1
        lr_t = lr * (K.sqrt(1. - K.pow(self.beta_2, t)) /
                     (1. - K.pow(self.beta_1, t)))

        ms = [K.zeros(K.int_shape(p), dtype=K.dtype(p)) for p in params]
        vs = [K.zeros(K.int_shape(p), dtype=K.dtype(p)) for p in params]
        if self.amsgrad:
            vhats = [K.zeros(K.int_shape(p), dtype=K.dtype(p)) for p in params]
        else:
            vhats = [K.zeros((1,)) for _ in params]
        self.weights = [self.iterations] + ms + vs + vhats

        for p, g, m, v, vhat in zip(params, grads, ms, vs, vhats):
            m_t = (self.beta_1 * m) + (1. - self.beta_1) * g
            v_t = (self.beta_2 * v) + (1. - self.beta_2) * K.square(g)
            if self.amsgrad:
                vhat_t = K.maximum(vhat, v_t)
                step = lr_t * m_t / (K.sqrt(vhat_t) + self.epsilon)
                self.updates.append(K.update(vhat, vhat_t))
            else:
                step = lr_t * m_t / (K.sqrt(v_t) + self.epsilon)

            self.updates.append(K.update(m, m_t))
            self.updates.append(K.update(v, v_t))
            
            if self.layca:
                new_p = layca(p, step, lr)
            else:
                new_p =  p - step

            # Apply constraints.
            if getattr(p, 'constraint', None) is not None:
                new_p = p.constraint(new_p)

            self.updates.append(K.update(p, new_p))
        return self.updates

    def get_config(self):
        config = {'lr': float(K.get_value(self.lr)),
                  'beta_1': float(K.get_value(self.beta_1)),
                  'beta_2': float(K.get_value(self.beta_2)),
                  'decay': float(K.get_value(self.decay)),
                  'epsilon': self.epsilon,
                  'amsgrad': self.amsgrad,
                  'layca':self.layca}
        base_config = super(Adam, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))

class Adagrad(Optimizer):
    """Adagrad optimizer.
    It is recommended to leave the parameters of this optimizer
    at their default values.
    # Arguments
        lr: float >= 0. Learning rate.
        epsilon: float >= 0. If `None`, defaults to `K.epsilon()`.
        decay: float >= 0. Learning rate decay over each update.
        layca: boolean, wether to apply layca or not
    # References
        - [Adaptive Subgradient Methods for Online Learning and Stochastic Optimization](http://www.jmlr.org/papers/volume12/duchi11a/duchi11a.pdf)
    """

    def __init__(self, lr=0.01, epsilon=None, decay=0., layca = False,**kwargs):
        super(Adagrad, self).__init__(**kwargs)
        with K.name_scope(self.__class__.__name__):
            self.lr = K.variable(lr, name='lr')
            self.decay = K.variable(decay, name='decay')
            self.iterations = K.variable(0, dtype='int64', name='iterations')
        if epsilon is None:
            epsilon = K.epsilon()
        self.epsilon = epsilon
        self.initial_decay = decay
        
        self.layca = layca

    @interfaces.legacy_get_updates_support
    def get_updates(self, loss, params):
        
        grads = self.get_gradients(loss, params)
        shapes = [K.int_shape(p) for p in params]
        accumulators = [K.zeros(shape) for shape in shapes]
        self.weights = accumulators
        self.updates = [K.update_add(self.iterations, 1)]

        lr = self.lr
        if self.initial_decay > 0:
            lr *= (1. / (1. + self.decay * K.cast(self.iterations,
                                                  K.dtype(self.decay))))

        for p, g, a in zip(params, grads, accumulators):
            new_a = a + K.square(g)  # update accumulator
            self.updates.append(K.update(a, new_a))
            step = lr * g / (K.sqrt(new_a) + self.epsilon)
            
            if self.layca:
                new_p = layca(p, step, lr)
            else:
                new_p =  p - step
            
            # Apply constraints.
            if getattr(p, 'constraint', None) is not None:
                new_p = p.constraint(new_p)

            self.updates.append(K.update(p, new_p))
        return self.updates

    def get_config(self):
        config = {'lr': float(K.get_value(self.lr)),
                  'decay': float(K.get_value(self.decay)),
                  'epsilon': self.epsilon,
                  'layca':self.layca}
        base_config = super(Adagrad, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))


layer_rotation_monitoring.py
'''
Methods for recording and plotting layer rotation curves
'''

import numpy as np
from scipy.spatial.distance import cosine

import matplotlib
import matplotlib.pyplot as plt

from keras.callbacks import Callback
import keras.backend as K
from keras.losses import categorical_crossentropy

def get_kernel_layer_names(model):
    '''
    collects name of all layers of a model that contain a kernel in topological order (input layers first).
    '''
    layer_names = []
    for l in model.layers:
        if len(l.weights) >0:
            if 'kernel' in l.weights[0].name:
                layer_names.append(l.name)
    return layer_names

def plot_layer_rotation_curves(distances, ax = None):
    '''
    utility to plot the layer-wise cosine distances between current parameters and initial parameters, 
        as measured over training (i.e. layer rotation curves).
    deviations is a list of lists with epoch index in first axis, layer index in second axis, 
        containing the cosine distances for each layer as recorded over training
    '''
    distances = np.array(distances)
    
    # get one color per layer
    cm = plt.get_cmap('viridis')
    cm_inputs = np.linspace(0,1,distances.shape[1])
    
    if not ax:
        ax = plt.subplot(1,1,1)
    for i in range(distances.shape[-1]):
        layer = i
        ax.plot(np.arange(distances.shape[0]+1), [0]+list(distances[:,layer]), label = str(layer), color = cm(cm_inputs[i]))

    ax.set_ylim([0,1.])
    ax.set_xlim([0,distances.shape[0]])
     
    ax.set_xlabel('Epoch')
    ax.set_ylabel('Cosine distance')

def compute_layer_rotation(current_model, initial_w):
    '''
    for each layer, computes cosine distance between current weights and initial weights
    initial_w is a list of tuples containing layer name and corresponding initial numpy weights
    '''
    s = []
    for l_name, w in initial_w:
        s.append(cosine( current_model.get_layer(l_name).get_weights()[0].flatten(), w.flatten()))
    return s

class LayerRotationCurves(Callback):
    '''
    Computes and saves layer rotation curves during training
    '''
    def __init__(self, batch_frequency=np.inf):
        '''
        batch_frequency is the frequency at which the cosine distances are computed (minimum once per epoch)
        '''
        super().__init__()
        self.batch_frequency = batch_frequency
        
        self.memory = []
    
    def set_model(self,model):
        super().set_model(model)
        layer_names = get_kernel_layer_names(model) 
        
        # initial_w is a list of tuples containing layer name and corresponding initial numpy weights
        self.initial_w = list(zip(layer_names,[model.get_layer(l).get_weights()[0] for l in layer_names]))
        
    def on_batch_end(self, batch, logs=None):
        if batch % self.batch_frequency == 0: #batch 0 is accepted, batch resets at 0 at every epoch

            dist = compute_layer_rotation(self.model, self.initial_w)

            self.memory.append(dist)
    
    def plot(self,ax = None):
        plot_layer_rotation_curves(self.memory,ax)


classification.py
import numpy as np
import sys
import random
import os
import tqdm
from PIL import ImageFile
from sklearn.metrics import classification_report, balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from keras.layers.core import Dense, Flatten, Dropout
from keras.layers.normalization import BatchNormalization
from keras.models import Model
from keras.preprocessing.image import ImageDataGenerator
from keras.callbacks import ModelCheckpoint, TensorBoard
from keras.preprocessing import image
from keras.models import model_from_json
from keras.callbacks import LearningRateScheduler

from layer_rotation_control import Adam
from layer_rotation_monitoring import LayerRotationCurves


# TODO fix this constants
ImageFile.LOAD_TRUNCATED_IMAGES = True

img_width, img_height = 224, 224
CATEGORIES = range(16)
CHECKPOINT_DIR = './checkpoints/'
checkpoint_path = os.path.join(CHECKPOINT_DIR,
                               "weights-{epoch:02d}-{val_acc:.2f}.hdf5")

# TODO get this numbers automatically
resnet_weights = 'resnet50_weights_tf_dim_ordering_tf_kernels_notop.h5'
nb_validation_samples = 1147
nb_train_samples = 13650


def train(arch_file="model.json", weights_file='model.h5',
          train_data_dir="train", validation_data_dir="sample",
          arch='nasnet', epochs=500, batch_size=16,
          tb_callback=True, checkpoints=True):

    input_shape = (img_width, img_height, 3)
    shape = (img_width, img_height)

    # # Choose network to use
    if arch == 'resnet':
        from keras.applications.resnet50 import ResNet50
        pretrained_model = ResNet50(include_top=False,
                                    input_shape=input_shape,
                                    weights='imagenet')  # resnet_weights)
    if arch == 'nasnet':
        from keras.applications.nasnet import NASNetLarge
        pretrained_model = NASNetLarge(input_shape=None,
                                       include_top=True,
                                       weights='imagenet',
                                       input_tensor=None,
                                       pooling=None,
                                       classes=1000)

    if pretrained_model.output.shape.ndims > 2:
        output = Flatten()(pretrained_model.output)
    else:
        output = pretrained_model.output

    output = BatchNormalization()(output)
    output = Dropout(0.5)(output)
    output = Dense(512, activation='relu')(output)
    output = BatchNormalization()(output)
    output = Dropout(0.5)(output)
    output = Dense(len(CATEGORIES), activation='softmax')(output)
    model = Model(pretrained_model.input, output)
    for layer in pretrained_model.layers:
        layer.trainable = False
    model.summary(line_length=200)
    lr = 0.001
    opt = Adam(lr, layca=True)

    # a keras callback to record layer rotation
    lrc = LayerRotationCurves()

    # learning rate schedule: divide learning rate by 5 at epochs 70 and 90
    def schedule(epoch):
        new_lr = lr
        if epoch > 70:
            new_lr *= 0.2
        if epoch > 90:
            new_lr *= 0.2
        return new_lr

    lrs = LearningRateScheduler(schedule)

    callbacks = [lrc, lrs]
    model.compile(loss='categorical_crossentropy',
                  optimizer=opt,
                  weighted_metrics=['accuracy'],
                  metrics=['accuracy'])

    # Save model arch to json
    model_json = model.to_json()
    with open(arch_file, "w") as json_file:
        json_file.write(model_json)

    # Data augmentation with generators
    _train_gen = ImageDataGenerator(preprocessing_function=preprocess_input,
                                    rotation_range=15,
                                    width_shift_range=.15,
                                    height_shift_range=.15,
                                    shear_range=0.15,
                                    zoom_range=0.15,
                                    channel_shift_range=1,
                                    horizontal_flip=True,
                                    vertical_flip=False)

    _test_gen = ImageDataGenerator(preprocessing_function=preprocess_input,
                                   rotation_range=15,
                                   width_shift_range=.15,
                                   height_shift_range=.15,
                                   shear_range=0.15,
                                   zoom_range=0.15,
                                   channel_shift_range=1,
                                   horizontal_flip=True,
                                   vertical_flip=False)

    train_generator = _train_gen.flow_from_directory(train_data_dir,
                                                     target_size=shape,
                                                     batch_size=batch_size,
                                                     class_mode='categorical')

    val_generator = _test_gen.flow_from_directory(validation_data_dir,
                                                  target_size=shape,
                                                  batch_size=batch_size,
                                                  class_mode='categorical')

    if tb_callback:
        tb_callback = TensorBoard(log_dir='./tb_callback',
                                  histogram_freq=0,
                                  write_graph=True,
                                  write_images=False)
        callbacks.append(tb_callback)

    if checkpoints:
        cp_callback = ModelCheckpoint(checkpoint_path,
                                      monitor='val_acc',
                                      save_weights_only=True,
                                      save_best_only=True,
                                      verbose=1)
        callbacks.append(cp_callback)
    model.fit_generator(train_generator,
                        steps_per_epoch=500,
                        validation_data=val_generator,
                        validation_steps=75,
                        epochs=epochs,
                        workers=4,
                        callbacks=callbacks)

    model.save_weights(weights_file)
    # plot layer rotation curves
    # plt.figure()
    # lrc.plot()


def preprocess_img(img_path):
    img_path = img_path
    img = image.load_img(img_path, target_size=(224, 224))
    img = image.img_to_array(img)
    img = np.expand_dims(img, axis=0)
    img = preprocess_input(img)
    return img
    # img = cv2.cvtColor(cv2.imread(img_path), cv2.COLOR_BGR2RGB)
    # #img = cv2.imread(img_path)
    # img = img / 255.
    # img = cv2.resize(img, (img_width, img_height))
    # if K.image_data_format() == 'channels_first':
    #     img = np.reshape(img, [1, 3, img_width, img_height])
    # else:
    #     img = np.reshape(img, [1, img_width, img_height, 3])
    # return img


def test(arch_file='model.json', weights_file='model.h5', test_dir='sample',
         arch='nasnet'):  #  TODO ARCH DEFINED BY ARCH FILE
    with open(arch_file, 'r') as json_file:
        loaded_model_json = json_file.read()
    loaded_model = model_from_json(loaded_model_json)
    if arch == 'nasnet':
        from keras.applications.nasnet import preprocess_input
    else:
        from keras.applications.resnet50 import preprocess_input
    loaded_model.load_weights(weights_file)
    y_true = []
    y_pred = []
    for _class in tqdm.tqdm(os.listdir(test_dir)):
        if _class == '.gitkeep':
            continue
        files = os.listdir(os.path.join(test_dir, _class))
        for _file in files:
            img = preprocess_img(os.path.join(test_dir, _class, _file))
            _predicted_class = np.argmax(loaded_model.predict(img))
            y_true.append(int(_class))
            y_pred.append(int(_predicted_class))
    report = classification_report(y_true, y_pred)
    score = balanced_accuracy_score(y_true, y_pred)
    print('Balanced score:', score)
    print(confusion_matrix(y_true, y_pred))
    return report


def get_random_sample_from_dataset(output_dir, dataset_dir='train'):
    if not os.path.exists(output_dir):
        os.mkdir(output_dir)
    for _class in os.listdir(dataset_dir):
        if not os.path.isdir(os.path.join(dataset_dir, _class)):
            continue
        files = os.listdir(os.path.join(dataset_dir, _class))
        dest_folder = os.path.join(output_dir, _class)
        if not os.path.exists(dest_folder):
            os.mkdir(dest_folder)
        random.shuffle(files)
        files = files[:int(.2 * len(files))]
        for file in files:
            from_file = os.path.join(dataset_dir, _class, file)
            to_file = os.path.join(output_dir, _class, file)
            print("Adding {}".format(from_file))
            os.system('mv {} {}'.format(from_file, to_file))


if __name__ == '__main__':
    if len(sys.argv) < 1:
        print("BAD ARGS")
    if sys.argv[1] == 'train':
        print("TRAIN")
        gpu_devs = os.getenv('CUDA_VISIBLE_DEVICES', None)
        print(train())
    if sys.argv[1] == 'test':
        print("TEST")
        print(test(weights_file=sys.argv[2]))
    if sys.argv[1] == 'make_sample':
        print("MAKING NEW SAMPLE")
        get_random_sample_from_dataset(output_dir=sys.argv[2])
